userspace/falco/falco_metrics.cpp

```diff
@@ -134,34 +137,77 @@ std::string falco_metrics::to_text(const falco::app::state& state)
 
 		auto now = std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::system_clock::now().time_since_epoch()).count();
 
-		static_metrics.push_back(libs_metrics_collector.new_metric("duration_sec",
+		additional_wrapper_metrics.emplace_back(libs_metrics_collector.new_metric("duration_sec",
 																	METRICS_V2_MISC,
 																	METRIC_VALUE_TYPE_U64,
 																	METRIC_VALUE_UNIT_TIME_S_COUNT,
 																	METRIC_VALUE_METRIC_TYPE_MONOTONIC,
 																	(uint64_t)((now - agent_info->start_ts_epoch) / ONE_SECOND_IN_NS)));
 
-		for (auto metrics: static_metrics)
+		for (auto metric: additional_wrapper_metrics)
+		{
+			prometheus_metrics_converter.convert_metric_to_unit_convention(metric);
+			prometheus_text += prometheus_metrics_converter.convert_metric_to_text_prometheus(metric, "falcosecurity", "falco");
+		}
+
+		// Falco metrics categories
+		//
+		// rules_counters_enabled
+		if(state.config->m_metrics_flags & METRICS_V2_RULE_COUNTERS)
 		{
-			prometheus_metrics_converter.convert_metric_to_unit_convention(metrics);
-			prometheus_text += prometheus_metrics_converter.convert_metric_to_text_prometheus(metrics, "falcosecurity", "falco");
+			const stats_manager& rule_stats_manager = state.engine->get_rule_stats_manager();
+			const indexed_vector<falco_rule>& rules = state.engine->get_rules();
+			auto metric = libs_metrics_collector.new_metric("rules.matches_total",
+																	METRICS_V2_RULE_COUNTERS,
+																	METRIC_VALUE_TYPE_U64,
+																	METRIC_VALUE_UNIT_COUNT,
+																	METRIC_VALUE_METRIC_TYPE_MONOTONIC,
+																	rule_stats_manager.get_total().load());
+
+			prometheus_metrics_converter.convert_metric_to_unit_convention(metric);
+			prometheus_text += prometheus_metrics_converter.convert_metric_to_text_prometheus(metric, "falcosecurity", "falco");
+			for (size_t i = 0; i < rule_stats_manager.get_by_rule_id().size(); i++)
+			{
+				auto rule = rules.at(i);
+				std::string rules_metric_name = "rules." + rule->name;
+				// Separate processing of rules counter metrics given we add extra tags
+				auto metric = libs_metrics_collector.new_metric(rules_metric_name.c_str(),
+																	METRICS_V2_RULE_COUNTERS,
+																	METRIC_VALUE_TYPE_U64,
+																	METRIC_VALUE_UNIT_COUNT,
+																	METRIC_VALUE_METRIC_TYPE_MONOTONIC,
+																	rule_stats_manager.get_by_rule_id()[i]->load());
```

Comment:
What about minimizing the number of calls to get_by_rule_id ?

```suggestion
			auto rules_by_id = rule_stats_manager.get_by_rule_id();
			for (size_t i = 0; i < rules_by_id.size(); i++)
			{
				auto rule = rules.at(i);
				std::string rules_metric_name = "rules." + rule->name;
				// Separate processing of rules counter metrics given we add extra tags
				auto metric = libs_metrics_collector.new_metric(rules_metric_name.c_str(),
																	METRICS_V2_RULE_COUNTERS,
																	METRIC_VALUE_TYPE_U64,
																	METRIC_VALUE_UNIT_COUNT,
																	METRIC_VALUE_METRIC_TYPE_MONOTONIC,
																	rules_by_id[i]->load());
```

Reply:
Incorporated, thanks.